presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3InputStream.java

```diff
@@ -0,0 +1,349 @@
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.net.HttpURLConnection.HTTP_BAD_REQUEST;
+import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.fs.FSExceptionMessages.CANNOT_SEEK_PAST_EOF;
+import static org.apache.hadoop.fs.FSExceptionMessages.NEGATIVE_SEEK;
+import static org.apache.hadoop.fs.FSExceptionMessages.STREAM_IS_CLOSED;
+
+public class PrestoS3InputStream
+        extends FSInputStream
+{
+    private static final DataSize MAX_SKIP_SIZE = new DataSize(1, MEGABYTE);
+    private static final int HTTP_RANGE_NOT_SATISFIABLE = 416;
+
+    private final AmazonS3 s3;
+    private final String host;
+    private final Path path;
+    private final int maxAttempts;
+    private final Duration maxBackoffTime;
+    private final Duration maxRetryTime;
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final PrestoS3FileSystemStats stats;
+
+    private InputStream in;
+    private long streamPosition;
+    private long nextReadPosition;
+
+    public PrestoS3InputStream(AmazonS3 s3, String host, Path path, int maxAttempts, Duration maxBackoffTime, Duration maxRetryTime, PrestoS3FileSystemStats stats)
+    {
+        this.s3 = requireNonNull(s3, "s3 is null");
+        this.host = requireNonNull(host, "host is null");
+        this.path = requireNonNull(path, "path is null");
+
+        checkArgument(maxAttempts >= 0, "maxAttempts cannot be negative");
+        this.maxAttempts = maxAttempts;
+        this.maxBackoffTime = requireNonNull(maxBackoffTime, "maxBackoffTime is null");
+        this.maxRetryTime = requireNonNull(maxRetryTime, "maxRetryTime is null");
+        this.stats = requireNonNull(stats, "stats is null");
+    }
+
+    @Override
+    public void close()
+    {
+        closed.set(true);
+        closeStream();
+    }
+
+    @Override
+    public int read(long position, byte[] buffer, int offset, int length)
+            throws IOException
+    {
+        checkClosed();
+        if (position < 0) {
+            throw new EOFException(NEGATIVE_SEEK);
+        }
+        checkPositionIndexes(offset, offset + length, buffer.length);
+        if (length == 0) {
+            return 0;
+        }
+
+        try {
+            return retry()
+                    .maxAttempts(maxAttempts)
+                    .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                    .stopOn(InterruptedException.class, PrestoS3FileSystem.UnrecoverableS3OperationException.class, EOFException.class, FileNotFoundException.class, AbortedException.class)
+                    .onRetry(stats::newGetObjectRetry)
+                    .run("getS3Object", () -> {
```

Comment:
I would move this lambda into a new method and use a method reference.

Reply:
updated